perf/pgbench.py

```diff
@@ -0,0 +1,128 @@
+        '''
+        Install Pgbench
+        '''
+        # check root privileges ($USER should be able to run commands
+        # as another user without requiring a password)
+        if os.geteuid() != 0:
+            self.cancel("This script requires root privileges, "
+                        "Please try again with 'sudo' or as 'root'")
+
+        # check if pgbench is installed
+        if (self.sm.check_installed("pgbench")):
+            self.log.info("Pgbench is already installed")
+        else:
+            pkg_mgr = self.get_pkg_manager()
+            if pkg_mgr:
+                self.install_packages(pkg_mgr)
+            else:
+                self.cancel("Unsupported Linux distribution")
+
+        # add current user to db
+        output = self.run_cmd(f"sudo -u postgres createuser --superuser {os.getlogin()}", True)
+        if output[0] and "already exists" not in output[1]:
+            self.cancel(f"Could not create role {os.getlogin()}")
+
+        # Setup the params
+        self.scaling_factor = self.params.get("scaling_factor", default=100)
+        self.worker_threads = self.params.get("worker_threads", default=cpu.online_count())
+        self.db_clients = self.params.get("db_clients", default=cpu.online_count())
+        self.protocol = self.params.get("protocol", default="prepared")
+        self.benchmark_duration = self.params.get("benchmark_duration", default=0)
+        self.transaction_count = self.params.get("transaction_count", default=0)
+
+        # validation: pgbench cannot accept both transaction count and duration
+        if self.benchmark_duration > 0 and self.transaction_count > 0:
+            self.log.warn("Pgbench cannot accept both max transaction count and duration of the test, proceeding with specified time limit")
+        elif self.benchmark_duration == 0 and self.transaction_count == 0:
+            self.log.warn("Transaction count / duration of the test not specified, defaulting to time limit of 120 secs")
+            self.benchmark_duration = 120
+
+        # create the db
+        self.run_cmd("createdb pgbench", False)
+
+        # setup tables
+        self.run_cmd(f"pgbench -i -s {self.scaling_factor} -n pgbench", False)
+
+    def test(self):
+
+        # run the benchmark
+        if self.benchmark_duration > 0:
+            result = self.run_cmd(f"pgbench --protocol={self.protocol} --jobs={self.worker_threads} --scale={self.scaling_factor} --client={self.db_clients}  -n --time={self.benchmark_duration} -r pgbench", False)
+        else:
+            result = self.run_cmd(f"pgbench --protocol={self.protocol} --jobs={self.worker_threads} --scale={self.scaling_factor} --client={self.db_clients}  -n --transactions={self.transaction_count} -r pgbench", False)
+        self.log.info("===== Pgbench benchmark results ======")
+        if not result[0]:
+            self.log.info(result[1])
+
+    def tearDown(self):
+        # destory the db
+        self.run_cmd("dropdb pgbench", True)
```

Comment:
Low Priority: I prefer tests to be idempotent, So _ideally_ you should also remove the packages you installed during tearDown()